xilem_core/src/state.rs

```diff
@@ -0,0 +1,219 @@
+/// the view value reads a value of type `T` when performing reconciliation and/or event handling.
+///
+/// This type implements [`ViewArgument`], and so any view operations for views which
+/// use this must be provided with a shared reference to `T` (i.e. `&T`).
+/// See [`Edit`] for the equivalent exclusive version of this.
+///
+/// # Examples
+///
+/// A simple component can display a single result:
+///
+/// ```rust
+/// # use xilem_core::docs::{DocsView as WidgetView, stateless_component};
+/// # use xilem_core::map_state;
+/// use xilem_core::Read;
+/// fn display_result() -> impl WidgetView<Read<f64>> {
+///     // ...
+/// #   map_state(stateless_component(), |_, ()| ())
+/// }
+/// ```
+///
+/// A component accepting only read access to a single value might not be that useful.
+/// Most components are likely to accept at least one [editable](Edit) parameter.
+/// This can be achieved using a tuple:
+///
+/// ```rust
+/// # use xilem_core::docs::{DocsView as WidgetView, stateless_component};
+/// # use xilem_core::map_state;
+/// use core::ops::Range;
+/// use xilem_core::{Read, Edit};
+/// fn slider() -> impl WidgetView<(Edit<f64>, Read<Range<f64>>)> {
+///     // ...
+/// #   map_state(stateless_component(), |(_result, _range), ()| ())
+/// }
+/// ```
+pub type Read<T> = &'static T;
+
+/// An alias to the "reference" form of a [`ViewArgument`].
+///
+/// This is used in `View` functions (instead of the expanded form) to decrease
+/// the amount of noise produced.
+// TODO: The name of this definitely needs rethinking.
+pub type Arg<'a, T> = <T as ViewArgument>::Params<'a>;
+
+// TODO: Do these need to be 'static?
+impl<T> ViewArgument for &'static T {
+    type Params<'a> = &'a T;
+    fn reborrow_mut<'input, 'a: 'input>(
+        params: &'input mut Self::Params<'a>,
+    ) -> Self::Params<'input> {
+        *params
+    }
+}
+
+impl<T> ViewArgument for &'static mut T {
+    type Params<'a> = &'a mut T;
+    fn reborrow_mut<'input, 'a: 'input>(
+        params: &'input mut Self::Params<'a>,
+    ) -> Self::Params<'input> {
+        &mut *params
+    }
+}
+
+impl ViewArgument for () {
+    type Params<'a> = ();
+    fn reborrow_mut<'input, 'a: 'input>((): &'input mut Self::Params<'a>) -> Self::Params<'input> {}
+}
+
+impl<T0: ViewArgument> ViewArgument for (T0,) {
+    type Params<'a> = (T0::Params<'a>,);
+    fn reborrow_mut<'input, 'a: 'input>(
+        (t0,): &'input mut Self::Params<'a>,
+    ) -> Self::Params<'input> {
+        (T0::reborrow_mut(t0),)
+    }
+}
+
+impl<T0: ViewArgument, T1: ViewArgument> ViewArgument for (T0, T1) {
+    type Params<'a> = (T0::Params<'a>, T1::Params<'a>);
+    fn reborrow_mut<'input, 'a: 'input>(
+        (t0, t1): &'input mut Self::Params<'a>,
+    ) -> Self::Params<'input> {
+        (T0::reborrow_mut(t0), T1::reborrow_mut(t1))
+    }
+}
+
+// TODO: Generate 2+ with a macro; maybe 1+, but then again for understandability purposes, having at least one
+// outside of the macro is appealing..
```

Comment:
I agree, the impl is short enough, that we can keep it for `(T0,)` and `(T0,T1)`. Should this happen in this PR?